spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -444,6 +446,39 @@ protected void handleMessageInternal(Message<?> message) throws Exception {
 		}
 	}
 
+	protected boolean isExpireGroupsUponCompletion() {
+		return false;
+	}
+
+	private void removeEmptyGroupAfterTimeout(MessageGroup messageGroup, long timeout) {
+		Object groupId = messageGroup.getGroupId();
+		UUID groupUuid = UUIDConverter.getUUID(groupId);
+		ScheduledFuture<?> scheduledFuture = getTaskScheduler()
+				.schedule(() -> {
+					Lock lock = this.lockRegistry.obtain(groupUuid.toString());
+
+					try {
+						lock.lockInterruptibly();
+						try {
+							this.messageStore.removeMessageGroup(groupId);
+						}
+						finally {
+							lock.unlock();
+						}
+					}
+					catch (InterruptedException e) {
+						Thread.currentThread().interrupt();
+						this.logger.debug("Thread was interrupted while trying to obtain lock");
+					}
+
+				}, new Date(System.currentTimeMillis() + timeout));
+
+		if (this.logger.isDebugEnabled()) {
+			this.logger.debug("Schedule empty MessageGroup [ " + groupId + "] for removal.");
+		}
+		this.expireGroupScheduledFutures.put(groupUuid, scheduledFuture);
```

Comment:
Debug message at line 407 may now not be correct - since it's not a `forceComplete` scheduled task.